include/boost/url/impl/decode.hpp

```diff
@@ -23,19 +23,15 @@
 namespace boost {
 namespace urls {
 
-template< class CharSet >
+template< class StringToken >
 auto
 decode(
-    string_view s,
-    decode_opts const& opt,
-    CharSet const& allowed) noexcept ->
-        result< decode_view >
+    BOOST_URL_PCT_STRING_VIEW s,
+    StringToken&& token,
+    decode_opts const& opt) noexcept ->
+        result< BOOST_URL_STRTOK_RETURN >
```

Comment:
hmm I don't think you can just put `BOOST_URL_STRTOK_RETURN` as a template param in `result`, but also this is not emitted in the docs since it is in impl/ so we don't need any of these macros

Reply:
![image](https://user-images.githubusercontent.com/5369819/192077904-725e54fb-df83-46fc-8cb3-47daa2d4b985.png)

Reply:
I don't like this at all, returning `result<StringToken::result_type>`, this is weird

Reply:
What do you suggest?

Reply:
Well, look at the function

```
template< class StringToken >
auto
decode(
    pct_string_view s,
    StringToken&& token,
    decode_opts const& opt) noexcept ->
        result< typename StringToken::result_type >
{
    return s.decode( opt, std::move(token) );
}
```

How can this possibly fail? `pct_string_view` is already validated, so failure is impossible. Also, `s.decode()` doesn't return `result<>` so I'm not sure what is going on here? Lets discuss...

Reply:
Oh... OK. I got a little confused by the pct_string_view being documented as string_view.